Initial PR

davemfish opened the pull request:
This PR includes a first draft of a module with a class for representing a Metadata Control File in memory, and with convenient methods for adding metadata.

There are also tests and an Actions workflow.

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+# It's not clear to me why 'spatial' is type 'array' instead of
+# 'object', since it contains 'properties'.
+MCF_SCHEMA['properties']['identification']['properties'][
+    'extents']['properties']['spatial']['type'] = 'object'
```
@phargogh I just discovered that overriding this bit of the MCF schema results in mcf docs that don't translate to XML using pygeometa's jinja templates. So I'll probably need to revert back to using `array` here and in a couple other places before we can write XML.

davemfish replied:
Yes, what I don't understand (though it seems to work) is why the jsonschema for an array (`spatial`) has "properties" instead of "items". From pygeometa's `core.yml`:

```yaml
            extents:
                type: object
                description: spatial and temporal extents
                properties:
                    spatial:
                        type: array
                        description: spatial extent and CRS
                        properties:
                            bbox:
                                type: array
                                description: bounding box of resource
                                items:
                                    type: number
                                minItems: 4
                                maxItems: 6
                            crs:
                                type: string
                                description: coordinate reference system of bbox
                                default: 'http://www.opengis.net/def/crs/OGC/1.3/CRS84'
                            description:
                                type: string
                                description: description of the geographic area using an identifier
                        required:
                            - bbox
```

phargogh replied:
Interesting!  I agree, that does seem like an oversight.  Looking at ISO19115, it appears that the extent is allowed to be defined N times, perhaps for N possible layers, in sorted order?  Even so, it appears that pygeometa's Jinja2 template only ever uses the first extent defined here, so maybe this isn't a commonly used part of the ISO spec.

phargogh reviewed (changes requested):
Thanks @davemfish !  As far as python OO goes, this is looking like a nice class.  I left a few comments where intent might be clarified, and where you could use python's data model to take advantage of some OO behavior details if you wanted to.  I think there was one other place where I think a method might make good sense to be private as well, but that's pretty minor.  The recursion makes a lot of sense to me and I'm glad we talked about it before hand.  Happy to talk more about anything else that would be helpful!

davemfish commented on pyproject.toml:
```diff
@@ -0,0 +1,38 @@
+dynamic = ["version", "dependencies", "optional-dependencies"]
+
+[build-system]
+requires = ["setuptools >= 40.6.0", "wheel"]
```
Yes, thank you!

davemfish commented on pyproject.toml:
```diff
@@ -0,0 +1,38 @@
+keywords = ["gis", "geospatial", "metadata"]
+classifiers = [
+    "Intended Audience :: Developers",
+    "Development Status :: 5 - Production/Stable",
```
Changed to "Development Status :: 2 - Pre-Alpha", which is one step beyond "Planning"

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+            for field in layer.schema:
+                attribute = {}
+                attribute['name'] = field.name
+                attribute['type'] = OGR_MCF_ATTR_TYPE_MAP[field.type]
```
Here's the schema for `attribute.type`:

```yaml
      type:
          type: string
          description: data type of attribute
          enum:
              - string
              - number
              - integer
              - object
              - array
              - boolean
```
`enum` allows for only one of those values. So I'm kind of leaning towards deleting the `attribute['type']` in that case -- it's not required by the schema for `attribute` -- rather than choosing an arbitrary value that might be incorrect.

phargogh replied:
I think that sounds like a good solution ... maybe we could issue a warning or something if we hit that case?

phargogh replied:
Patched in https://github.com/natcap/pygeometadata/pull/2/commits/9e78e012972017daa63b91249a57958e1ce5c90e

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+    def to_string(self):
+        pass
+
+    def get_spatial_info(self):
```
Yes, thanks for this suggestion. I also renamed it to `_set_spatial_info` since that is much more accurate.

I also changed the `__init__` logic to always call this method, even when an MCF already exists on disk, just in case the dataset was modified since the MCF was created.

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+        # we could use pygeometa.core.validate_mcf
+        jsonschema.validate(self.mcf, MCF_SCHEMA)
+
+    def to_string(self):
+        pass
```
Thanks for this tip! That was indeed the idea. It doesn't look like we detailed a use-case in the design doc though. I imagine it would be used for serialization, when that need arises.

An instance of an `MCF` includes a few attributes at the moment:
```python
        self.datasource = source_dataset_path
        self.mcf_path = f'{self.datasource}.yml'
        self.mcf = None  # the metadata dictionary
```

`a = MCF('data.tif')`

I could imagine that `str(a)`  would stringify all the attributes. But I could also imagine a use-case for just wanting to stringify `a.mcf`. So maybe `a.to_string()` would be an intuitive way to do that?

Anyway, I left this method empty because we don't have a use for it yet, so maybe I'll leave it like that for now?

phargogh replied:
Sounds good to me!  Yes, settling on the use case for that would be very wise.  To further add to the list of possibilities for what `__str__` could do, sometimes classes call `__repr__` instead, which could be useful in debugging, something like `<MCF at 0x12345abcdef>`.  Anyways, yes, let's defer that until later.

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+        if units is not None:
+            attribute['units'] = units
+
+        self.mcf['content_info']['attributes'].insert(idx, attribute)
```
Yes...I guess it would be equivalent to `self.mcf['content_info']['attributes'][idx] = attribute`

`attributes` is a list containing `attribute`. `attribute` is being modified in this method, and then being used to replace the pre-existing `attribute` at the same index. Am I missing your meaning?

I added to `test_raster_attributes` and `test_vector_attributes` to better illustrate the behavior I'm going for.

phargogh replied:
My main concern is about the usage of `insert`, which doesn't overwrite the element:

```python
>>> a = [1, 2, 3]
>>> a.insert(2, 'foo')
>>> a
[1, 2, 'foo', 3]
```

We can see this when adding an assertion to the test suite, where we're starting off with a 2-band raster (so 2 `attributes` sections), and the base attribute dict is being copied once per call to `insert`:

```python
>       self.assertEqual(len(mcf.mcf['content_info']['attributes']), 2)
E       AssertionError: 4 != 2
```

davemfish replied:
Ahhh, of course, thanks for spelling it out for me. Patched in e282ca3

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+        """
+        self.mcf['identification']['abstract'] = abstract
+
+    def keywords(self, keywords, section='default', keywords_type='theme',
```
Oh that's interesting. One concern is that our object is like,
```python
mcf = MCF(...)
print(mcf.mcf)  # the actual metadata dictionary
```
Since an instance of `MCF` has a couple other attributes besides the metadata dict. I wonder if there is a better way to design that, though. Would it be better if an instance of `MCF` was a direct representation of the MCF file?

Good advice on changing the method names, I'll do that.

phargogh replied:
>Since an instance of MCF has a couple other attributes besides the metadata dict. I wonder if there is a better way to design that, though. Would it be better if an instance of MCF was a direct representation of the MCF yaml?

Yep, I hear you on not wanting to allow access to these other attributes!  The usual python approach would be to make anything you don't want a user to access "private" by adding a `_` prefix.  [Here's a neat SO discussion](https://stackoverflow.com/a/7456865) about this in python and how attribute access compares to other languages.

>Would it be better if an instance of MCF was a direct representation of the MCF yaml?
As you said, I think the complex nested structure of the MCF yaml means that if we exclusively use that structure, then users of this library will need to be aware of the structure in order to use this library, which we don't want.  An interested informed user can always work directly with the upstream `pygeometa`, so I think we should be free to do what we think is right to make it easier to use and write spatial metadata.

davemfish commented:
Thanks for the feedback @phargogh . I think I addressed all your comments. There might be one or two things worth discussing further if it seems like we're not on the same page.

phargogh commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+import logging
+import os
+import yaml
+
```
Yeah, yaml is one of those things where it's _super_ common, but somehow not yet part of the stdlib.

Noted about the "commit suggestion" feature!

davemfish commented on src/pygeometadata/mcf.py:
```diff
@@ -0,0 +1,413 @@
+        if units is not None:
+            attribute['units'] = units
+
+        self.mcf['content_info']['attributes'].insert(idx, attribute)
```
Patched in e282ca3

phargogh reviewed (approved):
Thanks so much @davemfish !  This looks great, and congrats on a a first PR for this repo and class design!

The questions about how to represent the object and what to do about private attributes is tangential to this particular PR, so I'm happy to defer on those until later.
